sql3/planner/inbuiltfunctionsdate.go

```diff
@@ -196,3 +220,82 @@ func (n *callPlanExpression) EvaluateToTimestamp(currentRow []interface{}) (inte
 	//should we throw error or return nil if the conversion fails? what is the desired behaviour when ToTimestamp errors for one bad record in a batch of thousands?
 	return featurebase.ValToTimestamp(unit, num)
 }
+
+func (n *callPlanExpression) EvaluateDateTimeName(currentRow []interface{}) (interface{}, error) {
+	intervalEval, err := n.args[0].Evaluate(currentRow)
+	if err != nil {
+		return nil, err
+	}
+
+	dateEval, err := n.args[1].Evaluate(currentRow)
+	if err != nil {
+		return nil, err
+	}
+
+	// nil if anything is nil
+	if intervalEval == nil || dateEval == nil {
+		return nil, nil
+	}
+
+	//get the date value
+	coercedDate, err := coerceValue(n.args[1].Type(), parser.NewDataTypeTimestamp(), dateEval, parser.Pos{Line: 0, Column: 0})
+	if err != nil {
+		return nil, err
+	}
+
+	date, dateOk := coercedDate.(time.Time)
+	if !dateOk {
+		return nil, sql3.NewErrInternalf("unable to convert value")
+	}
+
+	//get the interval value
+	coercedInterval, err := coerceValue(n.args[0].Type(), parser.NewDataTypeString(), intervalEval, parser.Pos{Line: 0, Column: 0})
+	if err != nil {
+		return nil, err
+	}
+
+	interval, intervalOk := coercedInterval.(string)
+	if !intervalOk {
+		return nil, sql3.NewErrInternalf("unable to convert value")
+	}
+
+	switch strings.ToUpper(interval) {
+	case intervalYear:
+		return fmt.Sprint(date.Year()), nil
+
+	case intervalYearDay:
+		return fmt.Sprint(date.YearDay()), nil
+
+	case intervalMonth:
+		return fmt.Sprint(date.Month()), nil
+
+	case intervalDay:
+		return fmt.Sprint(date.Day()), nil
+
+	case intervalWeeKDay:
+		return fmt.Sprint(date.Weekday()), nil
+
+	case intervalWeek:
+		_, isoWeek := date.ISOWeek()
+		return fmt.Sprint(isoWeek), nil
+
+	case intervalHour:
+		return fmt.Sprint(date.Hour()), nil
+
+	case intervalMinute:
+		return fmt.Sprint(date.Minute()), nil
+
+	case intervalSecond:
+		return fmt.Sprint(date.Second()), nil
+
+	case intervalMillisecond:
+		return fmt.Sprint(date.Nanosecond() * 1000 * 1000), nil
```

Comment:
time.Nanosecond() represents the second offset in nano seconds, to extract milli, micro and nano seconds from this value the calculation needs to as follows, it was coded defective in the existing datepart() and I fixed it as part of the work I did for datetimeAdd() :
	case intervalMillisecond:
		return int64(date.Nanosecond() / 1000000), nil
	case intervalMicrosecond:
		return int64((date.Nanosecond() % 1000000) / 1000), nil
	case intervalNanosecond:
		return int64(date.Nanosecond() % 1000), nil

Reply:
I think that code answers the wrong question though? like if the time is 11:22:33.444555666, and the query asks for nanoseconds, I think datepart and datetimename are supposed to return the antire fractional seconds part, expressed as nanoseconds, so it'd want to be 444555666, not just 666?

Reply:
I think I've got it sorted out to what it probably should be now. MS gives 444, US gives 444555, and NS gives 444555666. Good catch on the part where MS needed to be / instead of * though, that went right by me.